R/selector.R

```diff
@@ -0,0 +1,235 @@
+      id <- str_remove(tmp_id, "^#")
+      selector <- str_remove(selector, tmp_id, fixed = TRUE)
+    }
+
+    classes <- str_remove(str_match_all(selector, "\\.[^.:[]+"), "^\\.")
+    if (length(classes) == 0) {
+      classes <- NULL
+    }
+    # if (!is.null(classes)) {
+    #   selector <- str_remove(selector, "\\.[^.:[]+")
+    # }
+  }
+
+  structure(class = "shiny_selector", list(
+    element = element,
+    id = id,
+    classes = classes,
+    type = type
+  ))
+}
+
+
+as_selector_list <- function(selector) {
+  selector <- as_selector(selector)
+  if (inherits(selector, "shiny_selector")) {
+    selector <- structure(class = "shiny_selector_list", list(selector))
+  }
+  if (length(selector) == 1) {
+    if (selector[[1]]$type == SELECTOR_CHILD) {
+      stop(
+        "Direct children selector, `>`, must not be the only element in a css selector.\n",
+        "Please add more selector information, such as `div > span`."
+      )
+    }
+  }
+
+
+  selector
+}
+
+#' @export
+format.shiny_selector <- function(x, ...) {
+  switch(x$type,
+    "everything" = "*",
+    "child" = ">",
+    paste0(x$element, if (!is.null(x$id)) paste0("#", x$id), if (!is.null(x$classes)) paste0(".", x$classes, collapse = ""))
+  )
+}
+#' @export
+format.shiny_selector_list <- function(x, ...) {
+  paste0(unlist(lapply(x, format, ...)), collapse = " ")
+}
+
+#' @export
+print.shiny_selector <- function(x, ...) {
+  cat("// htmltools css selector\n")
+  cat(format(x, ...), "\n")
+}
+#' @export
+print.shiny_selector_list <- function(x, ...) {
+  cat("// htmltools css selector list\n")
+  cat(format(x, ...), "\n")
+}
+
+
+
+str_replace <- function(text, pattern, replacement, fixed = FALSE) {
+  sub(pattern = pattern, replacement = replacement, x = text, perl = !fixed, fixed = fixed)
+}
+
+str_replace_all <- function(text, pattern, replacement, fixed = FALSE) {
+  gsub(pattern = pattern, replacement = replacement, x = text, perl = !fixed, fixed = fixed)
+}
+
+str_remove <- function(x, pattern, ...) {
+  str_replace(x, pattern, "", ...)
+}
+str_remove_all <- function(x, pattern, ...) {
+  str_replace_all(x, pattern, "", ...)
+}
+
+trim_leading <- function(text) {
+  str_remove_all(text, pattern = "^\\s+")
+}
+
+trim_trailing <- function(text) {
+  str_remove_all(text, pattern = "\\s+$")
+}
+
+str_trim <- function(text, side = "both") {
+  if (side == "both" || side == "left") {
+    text <- trim_leading(text)
+  }
+  if (side == "both" || side == "right") {
+    text <- trim_trailing(text)
+  }
+  text
+}
+
+str_detect <- function(text, pattern, fixed = FALSE) {
+  grepl(pattern = pattern, x = text, perl = !fixed, fixed = fixed)
+}
+
+# finds first, NOT all
+str_match_first <- function(x, pattern, ...) {
+  reg_info <- regexpr(pattern, x, ...)
+  if (length(reg_info) == 1 && reg_info == -1) {
+    return(NULL)
+  }
+
+  regmatches(x, reg_info)
+}
+
+# return a vector of matches or NULL
+str_match_all <- function(x, pattern, ...) {
+  stopifnot(length(x) == 1)
```

Comment:
Here and elsewhere, avoid `stopifnot` because it's slow.

```R
f1 <- function(x) {
  stopifnot(x > 0)
  NULL
}
f2 <- function(x) {
  if (!(x > 0)) stop()
  NULL
}
bench::mark(
  f1(1),
  f2(1)
)
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(1)        1.92µs    2.7µs   253774.        0B        0
#> 2 f2(1)         271ns    323ns  1475025.        0B        0
```